File: landing/src/services/backgroundRemoval.ts
import type * as ortTypes from 'onnxruntime-web';

import {
  BACKGROUND_MODEL_SIZE,
  canvasToDataUrl,
} from './imageProcessing';

const MODEL_URL = '/models/background_removal.onnx';
const MODEL_INPUT_NAME = 'input.1';
const MEAN = [0.485, 0.456, 0.406] as const;
const STD = [0.229, 0.224, 0.225] as const;

type OrtRuntime = typeof import('onnxruntime-web/wasm');

let runtimePromise: Promise<OrtRuntime> | undefined;
let sessionPromise: Promise<ortTypes.InferenceSession> | undefined;

/**
 * Browser equivalent of the former ONNX background-removal service.
 *
 * The model and ONNX WASM runtime are served by the website itself. No image
 * bytes leave the browser during segmentation.
 */
export class BackgroundRemovalService {
  async initialize(): Promise<void> {
    await this.getSession();
  }

  async removeBackground(source: HTMLCanvasElement): Promise<HTMLCanvasElement> {
    const [session, ort] = await Promise.all([this.getSession(), loadRuntime()]);
    const modelInput = createModelInput(source);
    const tensor = new ort.Tensor(
      'float32',
      modelInput.data,
      [1, 3, BACKGROUND_MODEL_SIZE, BACKGROUND_MODEL_SIZE],
    );

    const outputs = await session.run({ [MODEL_INPUT_NAME]: tensor });
    const output = outputs[session.outputNames[0] ?? ''];
    if (!output) {
      throw new Error('The background-removal model returned no mask.');
    }

    const mask = Array.from(output.data as Float32Array);
    return applyMask(source, mask, modelOutputWidth(output), modelOutputHeight(output));
  }

  private async getSession(): Promise<ortTypes.InferenceSession> {
    if (!sessionPromise) {
      sessionPromise = loadRuntime().then((ort) => {
        configureRuntime(ort);
        return ort.InferenceSession.create(MODEL_URL, {
          executionProviders: ['wasm'],
          graphOptimizationLevel: 'all',
        });
      }).catch((error: unknown) => {
        sessionPromise = undefined;
        throw new Error(`Background-removal model could not be loaded: ${errorMessage(error)}`);
      });
    }
    return sessionPromise;
  }
}

async function loadRuntime(): Promise<OrtRuntime> {
  if (!runtimePromise) {
    runtimePromise = import('onnxruntime-web/wasm').catch((error: unknown) => {
      runtimePromise = undefined;
      throw new Error(`Browser segmentation runtime could not be loaded: ${errorMessage(error)}`);
    });
  }
  return runtimePromise;
}

function configureRuntime(ort: OrtRuntime): void {
  // Keep the runtime compatible with a normal local static server. Threaded WASM
  // requires cross-origin isolation, so the single-threaded configuration is a
  // safer default for local development.
  ort.env.wasm.wasmPaths = '/models/';
  ort.env.wasm.numThreads = 1;
  ort.env.wasm.proxy = false;
}

function createModelInput(source: HTMLCanvasElement): { data: Float32Array } {
  const canvas = document.createElement('canvas');
  canvas.width = BACKGROUND_MODEL_SIZE;
  canvas.height = BACKGROUND_MODEL_SIZE;
  const context = canvas.getContext('2d');
  if (!context) {
    throw new Error('The browser could not create a segmentation canvas.');
  }
  context.imageSmoothingEnabled = true;
  context.imageSmoothingQuality = 'high';
  context.drawImage(source, 0, 0, BACKGROUND_MODEL_SIZE, BACKGROUND_MODEL_SIZE);

  const pixels = context.getImageData(0, 0, BACKGROUND_MODEL_SIZE, BACKGROUND_MODEL_SIZE).data;
  const pixelCount = BACKGROUND_MODEL_SIZE * BACKGROUND_MODEL_SIZE;
  const values = new Float32Array(pixelCount * 3);

  for (let index = 0; index < pixelCount; index += 1) {
    const rgbaIndex = index * 4;
    values[index] = normalize(pixels[rgbaIndex] ?? 0, MEAN[0], STD[0]);
    values[pixelCount + index] = normalize(pixels[rgbaIndex + 1] ?? 0, MEAN[1], STD[1]);
    values[pixelCount * 2 + index] = normalize(pixels[rgbaIndex + 2] ?? 0, MEAN[2], STD[2]);
  }

  return { data: values };
}

function normalize(value: number, mean: number, std: number): number {
  return (value / 255 - mean) / std;
}

function modelOutputWidth(output: ortTypes.Tensor): number {
  return output.dims.at(-1) ?? BACKGROUND_MODEL_SIZE;
}

function modelOutputHeight(output: ortTypes.Tensor): number {
  return output.dims.at(-2) ?? BACKGROUND_MODEL_SIZE;
}

function applyMask(
  source: HTMLCanvasElement,
  mask: number[],
  maskWidth: number,
  maskHeight: number,
): HTMLCanvasElement {
  const context = source.getContext('2d');
  if (!context) {
    throw new Error('The browser could not read the selected image.');
  }

  const image = context.getImageData(0, 0, source.width, source.height);
  const resizedMask = resizeMask(mask, maskWidth, maskHeight, source.width, source.height);
  const enhancedMask = enhanceMaskEdges(image.data, resizedMask, source.width, source.height);
  const smoothedMask = smoothMask(enhancedMask, source.width, source.height);

  for (let y = 0; y < source.height; y += 1) {
    for (let x = 0; x < source.width; x += 1) {
      const pixelIndex = y * source.width + x;
      const maskValue = smoothedMask[pixelIndex] ?? 0;
      const alpha = maskValue > 0.55
        ? 255
        : maskValue < 0.45
          ? 0
          : Math.round(((maskValue - 0.45) / 0.1) * 255);
      image.data[pixelIndex * 4 + 3] = alpha;
    }
  }

  const output = document.createElement('canvas');
  output.width = source.width;
  output.height = source.height;
  const outputContext = output.getContext('2d');
  if (!outputContext) {
    throw new Error('The browser could not create the processed image.');
  }
  outputContext.putImageData(image, 0, 0);
  // Calling toDataURL here validates that the browser can encode the same PNG
  // representation used by preview, inference, and history.
  canvasToDataUrl(output);
  return output;
}

function resizeMask(
  mask: number[],
  maskWidth: number,
  maskHeight: number,
  width: number,
  height: number,
): number[] {
  const resized = new Array<number>(width * height);
  for (let y = 0; y < height; y += 1) {
    const sourceY = (y * maskHeight) / height;
    const y1 = Math.min(maskHeight - 1, Math.floor(sourceY));
    const y2 = Math.min(maskHeight - 1, y1 + 1);
    const yWeight = sourceY - y1;

    for (let x = 0; x < width; x += 1) {
      const sourceX = (x * maskWidth) / width;
      const x1 = Math.min(maskWidth - 1, Math.floor(sourceX));
      const x2 = Math.min(maskWidth - 1, x1 + 1);
      const xWeight = sourceX - x1;
      const top = interpolate(mask[y1 * maskWidth + x1] ?? 0, mask[y1 * maskWidth + x2] ?? 0, xWeight);
      const bottom = interpolate(mask[y2 * maskWidth + x1] ?? 0, mask[y2 * maskWidth + x2] ?? 0, xWeight);
      resized[y * width + x] = interpolate(top, bottom, yWeight);
    }
  }
  return resized;
}

function interpolate(first: number, second: number, weight: number): number {
  return first * (1 - weight) + second * weight;
}

function enhanceMaskEdges(
  rgba: Uint8ClampedArray,
  mask: number[],
  width: number,
  height: number,
): number[] {
  const enhanced = [...mask];
  for (let y = 1; y < height - 1; y += 1) {
    for (let x = 1; x < width - 1; x += 1) {
      const left = (y * width + x - 1) * 4;
      const right = (y * width + x + 1) * 4;
      const up = ((y - 1) * width + x) * 4;
      const down = ((y + 1) * width + x) * 4;
      const gradientRed = Math.abs((rgba[right] ?? 0) - (rgba[left] ?? 0)) + Math.abs((rgba[down] ?? 0) - (rgba[up] ?? 0));
      const gradientGreen = Math.abs((rgba[right + 1] ?? 0) - (rgba[left + 1] ?? 0)) + Math.abs((rgba[down + 1] ?? 0) - (rgba[up + 1] ?? 0));
      const gradientBlue = Math.abs((rgba[right + 2] ?? 0) - (rgba[left + 2] ?? 0)) + Math.abs((rgba[down + 2] ?? 0) - (rgba[up + 2] ?? 0));
      const gradientMagnitude = (gradientRed + gradientGreen + gradientBlue) / 3;
      const current = mask[y * width + x] ?? 0;

      if (gradientMagnitude > 30 && current > 0.3 && current < 0.7) {
        let total = 0;
        let count = 0;
        for (let offsetY = -1; offsetY <= 1; offsetY += 1) {
          for (let offsetX = -1; offsetX <= 1; offsetX += 1) {
            total += mask[(y + offsetY) * width + x + offsetX] ?? 0;
            count += 1;
          }
        }
        const average = total / count;
        enhanced[y * width + x] = Math.max(0, Math.min(1, current + (average > 0.5 ? 0.1 : -0.1)));
      }
    }
  }
  return enhanced;
}

function smoothMask(mask: number[], width: number, height: number): number[] {
  const smoothed = new Array<number>(mask.length);
  for (let y = 0; y < height; y += 1) {
    for (let x = 0; x < width; x += 1) {
      let total = 0;
      let count = 0;
      for (let offsetY = -1; offsetY <= 1; offsetY += 1) {
        for (let offsetX = -1; offsetX <= 1; offsetX += 1) {
          const nextX = x + offsetX;
          const nextY = y + offsetY;
          if (nextX >= 0 && nextX < width && nextY >= 0 && nextY < height) {
            total += mask[nextY * width + nextX] ?? 0;
            count += 1;
          }
        }
      }
      smoothed[y * width + x] = count === 0 ? 0 : total / count;
    }
  }
  return smoothed;
}

function errorMessage(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}

File: background_removal.py
"""Local ONNX background removal for CropCheckUp input images.

The bundled model was recovered from the historical browser implementation.
This module deliberately imports NumPy, Pillow, and ONNX Runtime lazily so
importing the CLI does not initialize the segmentation runtime by itself.
"""

from __future__ import annotations

from pathlib import Path
from threading import Lock
from typing import Any, Callable

BACKGROUND_MODEL_SIZE = 320
MODEL_INPUT_NAME = "input.1"
MODEL_INPUT_SHAPE = (1, 3, BACKGROUND_MODEL_SIZE, BACKGROUND_MODEL_SIZE)
MEAN = (0.485, 0.456, 0.406)
STD = (0.229, 0.224, 0.225)


class BackgroundRemovalError(RuntimeError):
    """Raised when the background-removal model cannot process an image."""


SessionFactory = Callable[[Path], Any]


class BackgroundRemovalService:
    """Run the background-removal model, loading its session only on demand."""

    def __init__(self, model_path: Path, session_factory: SessionFactory | None = None):
        self.model_path = Path(model_path)
        self._session_factory = session_factory
        self._session: Any | None = None
        self._session_lock = Lock()
        self._input_name = MODEL_INPUT_NAME

    def initialize(self) -> None:
        """Load and validate the ONNX session without processing an image."""
        self._get_session()

    def remove_background(self, image: Any) -> Any:
        """Return an RGBA Pillow image with the model's foreground mask applied."""
        np = _load_numpy()
        session = self._get_session()

        try:
            source = image.convert("RGBA")
            model_input = create_model_input(source, np=np)
        except (AttributeError, OSError, ValueError) as error:
            raise BackgroundRemovalError(f"could not prepare image for background removal: {error}") from error

        try:
            outputs = session.run(None, {self._input_name: model_input})
        except Exception as error:  # ONNX Runtime exposes backend-specific exception types.
            raise BackgroundRemovalError(f"background-removal inference failed: {error}") from error

        mask = _extract_mask(outputs, np=np)
        return apply_background_mask(source, mask, np=np)

    def _get_session(self) -> Any:
        if self._session is None:
            with self._session_lock:
                if self._session is None:
                    self._session = self._create_session()
        return self._session

    def _create_session(self) -> Any:
        if not self.model_path.is_file():
            raise FileNotFoundError(f"background-removal model not found: {self.model_path}")

        if self._session_factory is not None:
            session = self._session_factory(self.model_path)
        else:
            try:
                import onnxruntime as ort
            except ImportError as error:
                raise RuntimeError(
                    "ONNX Runtime is required for background removal; "
                    "install the CLI dependencies with: python -m pip install -r requirements.txt"
                ) from error

            try:
                session = ort.InferenceSession(
                    str(self.model_path),
                    providers=["CPUExecutionProvider"],
                )
            except Exception as error:  # Keep the CLI error actionable across runtimes/platforms.
                raise BackgroundRemovalError(
                    f"could not load background-removal model {self.model_path}: {error}"
                ) from error

        self._validate_session(session)
        return session

    def _validate_session(self, session: Any) -> None:
        try:
            inputs = session.get_inputs()
            outputs = session.get_outputs()
        except AttributeError:
            # A small injected test double may only expose ``run``. Real ONNX
            # Runtime sessions always expose metadata and are validated below.
            return
        except Exception as error:
            raise BackgroundRemovalError(f"could not inspect background-removal model: {error}") from error

        if len(inputs) != 1:
            raise BackgroundRemovalError(
                f"background-removal model exposes {len(inputs)} inputs; expected one"
            )

        input_metadata = inputs[0]
        input_name = getattr(input_metadata, "name", None)
        if input_name != MODEL_INPUT_NAME:
            raise BackgroundRemovalError(
                f"background-removal model input is {input_name!r}; expected {MODEL_INPUT_NAME!r}"
            )
        self._input_name = input_name

        input_shape = _static_shape(getattr(input_metadata, "shape", None))
        if input_shape is not None and input_shape != MODEL_INPUT_SHAPE:
            raise BackgroundRemovalError(
                f"background-removal model input shape is {input_shape}; "
                f"expected {MODEL_INPUT_SHAPE}"
            )

        input_type = getattr(input_metadata, "type", None)
        if input_type is not None and input_type != "tensor(float)":
            raise BackgroundRemovalError(
                f"background-removal model input type is {input_type!r}; expected 'tensor(float)'"
            )

        if not outputs:
            raise BackgroundRemovalError("background-removal model exposes no mask output")

        output_shape = _static_shape(getattr(outputs[0], "shape", None))
        if output_shape is not None and (
            len(output_shape) != 4
            or output_shape[0] != 1
            or output_shape[1] != 1
            or output_shape[2] < 1
            or output_shape[3] < 1
        ):
            raise BackgroundRemovalError(
                f"background-removal mask output shape is {output_shape}; "
                "expected (1, 1, H, W)"
            )

        output_type = getattr(outputs[0], "type", None)
        if output_type is not None and output_type != "tensor(float)":
            raise BackgroundRemovalError(
                f"background-removal mask output type is {output_type!r}; expected 'tensor(float)'"
            )


def create_model_input(image: Any, *, np: Any | None = None) -> Any:
    """Create the normalized NCHW float32 tensor expected by the ONNX model."""
    np = np or _load_numpy()

    try:
        from PIL import Image
    except ImportError as error:
        raise RuntimeError(
            "Pillow is required for background removal; "
            "install the CLI dependencies with: python -m pip install -r requirements.txt"
        ) from error

    resized = image.convert("RGB").resize(
        (BACKGROUND_MODEL_SIZE, BACKGROUND_MODEL_SIZE),
        resample=Image.Resampling.BILINEAR,
    )
    pixels = np.asarray(resized, dtype=np.float32) / 255.0
    mean = np.asarray(MEAN, dtype=np.float32).reshape(1, 1, 3)
    std = np.asarray(STD, dtype=np.float32).reshape(1, 1, 3)
    normalized = (pixels - mean) / std
    return np.ascontiguousarray(np.transpose(normalized, (2, 0, 1))[None, ...], dtype=np.float32)


def apply_background_mask(image: Any, mask: Any, *, np: Any | None = None) -> Any:
    """Apply a model mask and return an RGBA Pillow image.

    The source alpha is multiplied by the new mask alpha so existing
    transparency is preserved instead of being treated as fully opaque.
    """
    np = np or _load_numpy()

    try:
        from PIL import Image
    except ImportError as error:
        raise RuntimeError(
            "Pillow is required for background removal; "
            "install the CLI dependencies with: python -m pip install -r requirements.txt"
        ) from error

    source = image.convert("RGBA")
    rgba = np.array(source, dtype=np.uint8, copy=True)
    if rgba.ndim != 3 or rgba.shape[2] != 4:
        raise BackgroundRemovalError(f"source image has invalid RGBA shape {rgba.shape}")

    mask_array = np.asarray(mask, dtype=np.float32)
    if mask_array.ndim != 2 or mask_array.size == 0:
        raise BackgroundRemovalError(f"background-removal mask has invalid shape {mask_array.shape}")
    if not np.isfinite(mask_array).all():
        raise BackgroundRemovalError("background-removal mask contains non-finite values")
    if float(mask_array.min()) < 0.0 or float(mask_array.max()) > 1.0:
        raise BackgroundRemovalError("background-removal mask values must be between 0 and 1")

    height, width = rgba.shape[:2]
    resized_mask = _resize_mask(mask_array, width, height, np=np)
    enhanced_mask = _enhance_mask_edges(rgba, resized_mask, np=np)
    smoothed_mask = _smooth_mask(enhanced_mask, np=np)

    mask_alpha = np.where(
        smoothed_mask > 0.55,
        255.0,
        np.where(smoothed_mask < 0.45, 0.0, ((smoothed_mask - 0.45) / 0.1) * 255.0),
    )
    mask_alpha = np.clip(np.rint(mask_alpha), 0, 255).astype(np.uint8)
    rgba[:, :, 3] = np.rint(mask_alpha.astype(np.float32) * rgba[:, :, 3] / 255.0).astype(np.uint8)
    return Image.fromarray(rgba)


def remove_background(image: Any, model_path: Path) -> Any:
    """Apply background removal using the cached service for ``model_path``."""
    return get_background_removal_service(model_path).remove_background(image)


_SERVICE_CACHE: dict[Path, BackgroundRemovalService] = {}
_SERVICE_CACHE_LOCK = Lock()


def get_background_removal_service(model_path: Path) -> BackgroundRemovalService:
    """Return one lazily initialized service per model path."""
    key = Path(model_path).expanduser().resolve()
    with _SERVICE_CACHE_LOCK:
        service = _SERVICE_CACHE.get(key)
        if service is None:
            service = BackgroundRemovalService(key)
            _SERVICE_CACHE[key] = service
        return service


def _extract_mask(outputs: Any, *, np: Any) -> Any:
    if outputs is None:
        raise BackgroundRemovalError("background-removal model returned no outputs")

    try:
        first_output = outputs[0]
    except (IndexError, KeyError, TypeError) as error:
        raise BackgroundRemovalError("background-removal model returned no mask output") from error

    try:
        mask = np.asarray(first_output)
    except (TypeError, ValueError, OverflowError) as error:
        raise BackgroundRemovalError(f"background-removal mask output is invalid: {error}") from error
    if mask.dtype.kind != "f":
        raise BackgroundRemovalError(
            f"background-removal mask output has type {mask.dtype}; expected floating-point values"
        )
    mask = np.asarray(mask, dtype=np.float32)
    if mask.ndim == 4 and mask.shape[0] == 1 and mask.shape[1] == 1:
        mask = mask[0, 0]
    elif mask.ndim == 3 and mask.shape[0] == 1:
        mask = mask[0]
    elif mask.ndim != 2:
        raise BackgroundRemovalError(
            f"background-removal mask has shape {mask.shape}; expected (H, W), (1, H, W), or (1, 1, H, W)"
        )

    if mask.size == 0:
        raise BackgroundRemovalError("background-removal mask is empty")
    if not np.isfinite(mask).all():
        raise BackgroundRemovalError("background-removal mask contains non-finite values")
    if float(mask.min()) < 0.0 or float(mask.max()) > 1.0:
        raise BackgroundRemovalError("background-removal mask values must be between 0 and 1")
    return np.ascontiguousarray(mask, dtype=np.float32)


def _resize_mask(mask: Any, width: int, height: int, *, np: Any) -> Any:
    mask_height, mask_width = mask.shape
    if (mask_width, mask_height) == (width, height):
        return mask.copy()

    source_y = np.arange(height, dtype=np.float32) * mask_height / height
    source_x = np.arange(width, dtype=np.float32) * mask_width / width
    y1 = np.floor(source_y).astype(np.intp)
    x1 = np.floor(source_x).astype(np.intp)
    y2 = np.minimum(mask_height - 1, y1 + 1)
    x2 = np.minimum(mask_width - 1, x1 + 1)
    y_weight = source_y - y1
    x_weight = source_x - x1

    top_left = mask[y1[:, None], x1[None, :]]
    top_right = mask[y1[:, None], x2[None, :]]
    bottom_left = mask[y2[:, None], x1[None, :]]
    bottom_right = mask[y2[:, None], x2[None, :]]
    top = top_left * (1.0 - x_weight[None, :]) + top_right * x_weight[None, :]
    bottom = bottom_left * (1.0 - x_weight[None, :]) + bottom_right * x_weight[None, :]
    return top * (1.0 - y_weight[:, None]) + bottom * y_weight[:, None]


def _enhance_mask_edges(rgba: Any, mask: Any, *, np: Any) -> Any:
    enhanced = mask.copy()
    height, width = mask.shape
    if height < 3 or width < 3:
        return enhanced

    pixels = rgba[:, :, :3].astype(np.int16, copy=False)
    gradient_red = np.abs(pixels[1:-1, 2:, 0] - pixels[1:-1, :-2, 0]) + np.abs(
        pixels[2:, 1:-1, 0] - pixels[:-2, 1:-1, 0]
    )
    gradient_green = np.abs(pixels[1:-1, 2:, 1] - pixels[1:-1, :-2, 1]) + np.abs(
        pixels[2:, 1:-1, 1] - pixels[:-2, 1:-1, 1]
    )
    gradient_blue = np.abs(pixels[1:-1, 2:, 2] - pixels[1:-1, :-2, 2]) + np.abs(
        pixels[2:, 1:-1, 2] - pixels[:-2, 1:-1, 2]
    )
    gradient_magnitude = (gradient_red + gradient_green + gradient_blue) / 3.0

    current = mask[1:-1, 1:-1]
    neighborhood = (
        mask[:-2, :-2]
        + mask[:-2, 1:-1]
        + mask[:-2, 2:]
        + mask[1:-1, :-2]
        + mask[1:-1, 1:-1]
        + mask[1:-1, 2:]
        + mask[2:, :-2]
        + mask[2:, 1:-1]
        + mask[2:, 2:]
    ) / 9.0
    condition = (gradient_magnitude > 30.0) & (current > 0.3) & (current < 0.7)
    adjusted = np.clip(current + np.where(neighborhood > 0.5, 0.1, -0.1), 0.0, 1.0)
    enhanced[1:-1, 1:-1] = np.where(condition, adjusted, current)
    return enhanced


def _smooth_mask(mask: Any, *, np: Any) -> Any:
    padded = np.pad(mask, 1, mode="constant", constant_values=0.0)
    padded_counts = np.pad(np.ones_like(mask), 1, mode="constant", constant_values=0.0)
    total = np.zeros_like(mask)
    count = np.zeros_like(mask)
    for offset_y in range(3):
        for offset_x in range(3):
            total += padded[offset_y : offset_y + mask.shape[0], offset_x : offset_x + mask.shape[1]]
            count += padded_counts[offset_y : offset_y + mask.shape[0], offset_x : offset_x + mask.shape[1]]
    return np.divide(total, count, out=np.zeros_like(total), where=count != 0)


def _static_shape(shape: Any) -> tuple[int, ...] | None:
    if not isinstance(shape, (list, tuple)) or any(not isinstance(value, int) for value in shape):
        return None
    return tuple(shape)


def _load_numpy() -> Any:
    try:
        import numpy as np
    except ImportError as error:
        raise RuntimeError(
            "NumPy is required for background removal; "
            "install the CLI dependencies with: python -m pip install -r requirements.txt"
        ) from error
    return np
